### ISSUE TITLE
Request: Native Vectorized Batch Evaluation for Parameterized Quantum Circuits (Similar to PennyLane's Parameter Broadcasting)

### ISSUE THREAD

**Issue by TaichiMatsubara** [NVIDIA/cuda-quantum#3664]
### Required prerequisites

- [x] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to check if your feature has already been mentioned or rejected in other issues.

### Describe the feature

I am the author of QTFPred, a hybrid quantum-classical machine learning model for genomic sequence analysis published in *Briefings in Bioinformatics* (2025). I am writing to request a feature that would enable efficient batch evaluation of parameterized quantum circuits in CUDA-Q.

[https://doi.org/10.1093/bib/bbaf604](url)

In QTFPred's quantum convolutional layer, the same quantum circuit with the same trainable parameters needs to be evaluated thousands of times per forward pass, with only the input data differing between evaluations. This pattern is fundamental to quantum convolution operations and is well-suited for GPU parallelization.

Currently, QTFPred is implemented using PennyLane, which provides this functionality through its **parameter broadcasting** feature. However, training is slow due to CPU-GPU data transfer overhead in the PennyLane/PyTorch architecture. I am exploring CUDA-Q as a potential solution because of its promise of GPU-native execution and JIT compilation, but I have found that similar batch execution capability does not appear to be available in CUDA-Q.

I believe adding this feature would greatly benefit the hybrid quantum-classical machine learning community.

---

### The Sliding Window Pattern

Quantum convolutional layers, like their classical counterparts, operate using a sliding window approach. For a 1D quantum convolution:

- **Input sequence**: A DNA sequence of length 1,001 base pairs, encoded as a tensor of shape (batch, 4 channels, 1001 positions)
- **Kernel size**: 4 (matching the 4 nucleotide channels: A, T, G, C)
- **Output**: For each sliding window position, the quantum circuit processes the local patch and produces expectation values

The key insight is that **every window position uses the exact same quantum circuit with the exact same trainable parameters**. The only difference between evaluations is the input data (which window of the sequence is being processed).

### Calculation of Circuit Evaluations per Forward Pass

For a single sample with sequence length 1,001 and kernel size 4:
- Number of window positions: 1,001 - 4 + 1 = 998
- Number of input channels: 4
- **Total circuit evaluations: 998 × 4 = 3,992 per sample**

For a training batch of 32 samples, this becomes 127,744 circuit evaluations per forward pass. During backpropagation with parameter-shift rules or SPSA, this number doubles or more.

### Why This Pattern is Ideal for GPU Parallelization

This computational pattern has several characteristics that make it perfectly suited for GPU acceleration:

1. **Identical circuit structure**: All evaluations use the same sequence of quantum gates
2. **Identical trainable parameters**: The same weight values are used across all evaluations
3. **Independent computations**: Each circuit evaluation is completely independent of the others
4. **Uniform memory access patterns**: State vectors can be laid out contiguously in GPU memory

These characteristics align perfectly with GPU SIMD (Single Instruction, Multiple Data) architecture, where the same operation can be applied to many data elements simultaneously.

---

## How PennyLane Handles This: Parameter Broadcasting

PennyLane provides seamless batch processing through its **parameter broadcasting** feature, which is documented in the QTFPred paper's Supplementary Method 3.

### Official PennyLane Resources

1. **Blog Post**: [PennyLane v0.24 - Parameter Broadcasting](https://pennylane.ai/blog/2022/06/pennylane-v024-released/#support-for-quantum-parameter-broadcasting-)
2. **Documentation**: [qml.QNode API](https://docs.pennylane.ai/en/stable/code/api/pennylane.QNode.html)
3. **Discussion**: [TorchLayer Broadcasting Thread](https://discuss.pennylane.ai/t/is-parameter-broadcasting-of-qml-qnn-torchlayer-available-in-the-inherited-class/3133/4)

### How Parameter Broadcasting Works

From the QTFPred Supplementary Method 3:

> "PennyLane's vectorization capabilities allow the simultaneous evaluation of the same quantum circuit across multiple input data points with shared parameters, drastically reducing computational overhead."

**Key Concept**:
- Instead of calling the quantum circuit 127,744 times in a Python loop
- Call it **once** with a batched input tensor of shape `(127744, 4)`
- PennyLane internally vectorizes the computation
- Returns batched output of shape `(127744, 4)`

**Mathematical Formulation**:

Traditional approach (inefficient):
```
for i = 1 to N:
    y_i = f(x_i; θ)
```

Broadcasting approach (efficient):
```
Y = f(X; θ)    # Single vectorized call
```

Where X ∈ ℝ^(N×d) and Y ∈ ℝ^(N×n) are computed in parallel.

### Development Environment

```
Operating System: Ubuntu 22.04 LTS
Python Version: 3.11.5
PyTorch Version: 2.1.0+cu118
PennyLane Version: 0.32.0
NumPy Version: 1.26.0
```

### Complete Working Example

```python
#!/usr/bin/env python3
"""
PennyLane Quantum Convolution Layer with Automatic Batch Processing (Parameter Broadcasting)

This demonstrates how PennyLane handles multiple circuit evaluations
efficiently through parameter broadcasting, as used in QTFPred.

Reference: QTFPred Supplementary Method 3
PennyLane Broadcasting: https://pennylane.ai/blog/2022/06/pennylane-v024-released/#support-for-quantum-parameter-broadcasting-
"""

import torch
import torch.nn as nn
import pennylane as qml
from pennylane import numpy as np
import time

# Configuration
N_QUBITS = 4      # Number of qubits (matches DNA encoding channels)
N_LAYERS = 1      # Number of entangling layers
SEQ_LEN = 101     # Sequence length for demonstration
KERNEL_SIZE = 4   # Convolution kernel size


def circuit_structured_multi(inputs, weight_entangle1, weight_entangle2, weight_rot1):
    """
    Data re-uploading quantum circuit with strongly entangling layers.

    This is the circuit structure used in QTFPred:
    1. Encodes input data as RY rotations
    2. Applies parameterized entangling layer
    3. Re-encodes input data (data re-uploading)
    4. Applies second entangling layer
    5. Re-encodes input data again
    6. Applies final rotation layer
    7. Measures expectation values of Pauli-Z on each qubit
    """
    n_inputs = inputs.shape[-1] if len(inputs.shape) > 1 else len(inputs)
    inputs_scaled = inputs * np.pi

    # First encoding
    qml.AngleEmbedding(inputs_scaled, wires=range(n_inputs), rotation='Y')

    # First entangling layer
    qml.StronglyEntanglingLayers(weights=weight_entangle1, wires=range(n_inputs))

    # Second encoding (data re-uploading)
    qml.AngleEmbedding(inputs_scaled, wires=range(n_inputs), rotation='Y')

    # Second entangling layer
    qml.StronglyEntanglingLayers(weights=weight_entangle2, wires=range(n_inputs))

    # Third encoding
    qml.AngleEmbedding(inputs_scaled, wires=range(n_inputs), rotation='Y')

    # Final rotation layer
    qml.broadcast(qml.Rot, wires=range(n_inputs), pattern="single", parameters=weight_rot1)

    # Return expectation value of Z on each qubit
    return [qml.expval(qml.PauliZ(i)) for i in range(n_inputs)]


class QConv1D_PennyLane(nn.Module):
    """
    Quantum Convolutional Layer using PennyLane with Parameter Broadcasting.

    Key feature: When forward() receives input with shape (batch, channels, length),
    it processes ALL windows across ALL batch elements efficiently through
    PennyLane's automatic parameter broadcasting - NOT a Python for-loop.

    Without broadcasting, QTFPred training would be impossible (hours per batch).
    With broadcasting, training is practical (seconds per batch).
    """

    def __init__(self, kernel_size=4, n_layers=1):
        super().__init__()
        self.kernel_size = kernel_size

        # Create quantum device
        self.dev = qml.device("default.qubit", wires=kernel_size)

        # Define weight shapes for the circuit
        weight_shapes = {
            "weight_entangle1": (n_layers, kernel_size, 3),
            "weight_entangle2": (n_layers, kernel_size, 3),
            "weight_rot1": (kernel_size, 3)
        }

        # Create QNode with PyTorch interface
        circuit = qml.QNode(
            circuit_structured_multi,
            self.dev,
            interface='torch',
            diff_method='best'
        )

        # Wrap as TorchLayer - this enables automatic parameter broadcasting!
        self.qconv = qml.qnn.TorchLayer(circuit, weight_shapes)

    def forward(self, x):
        """
        Forward pass with sliding window convolution.

        Input: x of shape (batch_size, in_channels, sequence_length)
        Output: shape (batch_size, in_channels * kernel_size, output_length)

        The key is that all windows are flattened into a single batch dimension,
        then processed with a SINGLE call to the quantum circuit (broadcasting).
        """
        batch_size, in_channels, input_width = x.size()
        output_width = input_width - self.kernel_size + 1

        # Extract sliding windows
        # Shape: (batch, output_width, channels, kernel_size)
        unfolded = x.unfold(2, self.kernel_size, 1).permute(0, 2, 1, 3)

        # Flatten for quantum processing (enables broadcasting)
        # Shape: (batch * output_width * channels, kernel_size)
        unfolded_flat = unfolded.reshape(-1, self.kernel_size)

        # === PARAMETER BROADCASTING ===
        # PennyLane's TorchLayer automatically handles this batch!
        # All circuits are evaluated efficiently without explicit Python loops.
        # This is the key feature that makes quantum convolution practical.
        qconv_output = self.qconv(unfolded_flat)

        # Reshape output back to original structure
        output = qconv_output.view(batch_size, output_width, in_channels, self.kernel_size)
        output = output.permute(0, 2, 3, 1).reshape(batch_size, -1, output_width)

        return output


def main():
    print("=" * 70)
    print("PennyLane Quantum Convolution - Parameter Broadcasting Demonstration")
    print("=" * 70)
    print("\nThis demonstrates the parameter broadcasting feature used in QTFPred.")
    print("Reference: QTFPred Supplementary Method 3")
    print("PennyLane Broadcasting: https://pennylane.ai/blog/2022/06/pennylane-v024-released/")

    # Create model
    model = QConv1D_PennyLane(kernel_size=KERNEL_SIZE, n_layers=N_LAYERS)

    # Create input data (simulating DNA sequence encoding)
    batch_size = 2
    x = torch.randn(batch_size, 4, SEQ_LEN)

    # Calculate expected circuit evaluations
    output_width = SEQ_LEN - KERNEL_SIZE + 1
    total_circuits = batch_size * output_width * 4

    print(f"\nConfiguration:")
    print(f"  Sequence length: {SEQ_LEN}")
    print(f"  Kernel size: {KERNEL_SIZE}")
    print(f"  Batch size: {batch_size}")
    print(f"  Output width: {output_width}")
    print(f"  Total circuit evaluations: {total_circuits}")

    # Forward pass with timing
    print(f"\nRunning forward pass...")
    start = time.time()
    output = model(x)
    elapsed = time.time() - start

    print(f"\nResults:")
    print(f"  Input shape: {x.shape}")
    print(f"  Output shape: {output.shape}")
    print(f"  Time: {elapsed:.2f} seconds")
    print(f"  Throughput: {total_circuits / elapsed:.1f} circuits/sec")

    print("\n" + "=" * 70)
    print(f"Key insight: PennyLane processed all {total_circuits} circuit evaluations")
    print("efficiently through parameter broadcasting, NOT Python loops.")
    print("\nWithout broadcasting: QTFPred training would be IMPOSSIBLE (hours per batch)")
    print("With broadcasting: QTFPred training is PRACTICAL (seconds per batch)")
    print("=" * 70)


if __name__ == "__main__":
    main()
```

### Sample Output

```
======================================================================
PennyLane Quantum Convolution - Parameter Broadcasting Demonstration
======================================================================

This demonstrates the parameter broadcasting feature used in QTFPred.
Reference: QTFPred Supplementary Method 3
PennyLane Broadcasting: https://pennylane.ai/blog/2022/06/pennylane-v024-released/

Configuration:
  Sequence length: 101
  Kernel size: 4
  Batch size: 2
  Output width: 98
  Total circuit evaluations: 784

Running forward pass...

Results:
  Input shape: torch.Size([2, 4, 101])
  Output shape: torch.Size([2, 16, 98])
  Time: 0.45 seconds
  Throughput: 1742.2 circuits/sec

======================================================================
Key insight: PennyLane processed all 784 circuit evaluations
efficiently through parameter broadcasting, NOT Python loops.

Without broadcasting: QTFPred training would be IMPOSSIBLE (hours per batch)
With broadcasting: QTFPred training is PRACTICAL (seconds per batch)
======================================================================
```

---

### Investigation of observe.py

I examined the CUDA-Q source code to understand how batch execution is currently handled. In `python/cudaq/runtime/observe.py`, I found the `__broadcastObserve` function:

```python
def __broadcastObserve(kernel, spin_operator, *args, shots_count=0):
    argSet = __createArgumentSet(*args)
    N = len(argSet)
    results = []
    for i, a in enumerate(argSet):  # <-- Python for-loop
        ctx = cudaq_runtime.ExecutionContext('observe', shots_count)
        ctx.totalIterations = N
        ctx.batchIteration = i
        ctx.setSpinOperator(spin_operator)
        cudaq_runtime.setExecutionContext(ctx)
        kernel(*a)  # <-- Each circuit executed sequentially
        res = ctx.result
        cudaq_runtime.resetExecutionContext()
        results.append(...)
    return results
```

This implementation processes each circuit evaluation sequentially using a Python loop, even when a 2D parameter array is passed. The `observe_async` variant provides asynchronous execution but does not fundamentally change this sequential pattern.

---

## Why I Am Looking to CUDA-Q

Despite the training speed limitations, I have high hopes for CUDA-Q as a potential solution. Based on my analysis, CUDA-Q offers several architectural advantages that could address the performance bottleneck:

| Aspect | PennyLane + PyTorch | CUDA-Q (Expected) |
|--------|---------------------|-------------------|
| **Execution Location** | CPU ↔ GPU data transfer required | GPU memory-native execution |
| **Compilation** | Python interpreter overhead | JIT compilation to C++ binaries |
| **Batch Processing** | Performance degrades with large batches | Optimized for GPU batch parallelism |

The combination of GPU-native execution, JIT compilation, and the potential for optimized batch processing makes CUDA-Q a compelling platform for hybrid quantum-classical machine learning. If vectorized batch execution were available, it would enable significant performance improvements for models like QTFPred.

---

## Feature Request

I would like to respectfully request consideration of the following enhancement:

**When `cudaq.observe` receives a 2D parameter array (indicating batch evaluation), could it utilize vectorized GPU operations internally rather than iterating through a Python loop?**

This would:
1. Enable efficient hybrid quantum-classical neural network training with CUDA-Q
2. Leverage CUDA-Q's GPU-native architecture for maximum performance
3. Maintain the simple, high-level API that CUDA-Q users appreciate
4. Provide performance competitive with other quantum ML frameworks

---

## Use Case Summary

- **Application**: Quantum Convolutional Neural Network for DNA sequence analysis
- **Model**: QTFPred - Quantitative Transcription Factor binding Prediction
- **Publication**: Briefings in Bioinformatics, 2025
- **Workload**: 3,992+ circuit evaluations per sample per forward pass
- **Training requirement**: Thousands of training iterations with gradient computation
- **Current challenge**: PennyLane's CPU-GPU transfer overhead limits training speed
- **Hope for CUDA-Q**: GPU-native execution with vectorized batch processing

---

## Environment

- CUDA-Q version: 0.13.0
- Target: nvidia
- GPU: NVIDIA RTX 6000 Ada Generation (48GB VRAM)
- Python: 3.11
- CUDA: 12.x

---

## Related Discussions

I found these related discussions that suggest others in the community have similar needs:

- [#1269 - hybrid computing with batch](https://github.com/NVIDIA/cuda-quantum/discussions/1269)
- [#62 - observe_n RFC](https://github.com/NVIDIA/cuda-quantum/issues/62)
- [#1486 - Sampling and broadcasting API improvements](https://github.com/NVIDIA/cuda-quantum/issues/1486)

---

## Closing

Thank you for considering this feature request. As the developer of QTFPred, I have firsthand experience with the challenges of training hybrid quantum-classical models, and I believe that efficient batch circuit evaluation would significantly enhance CUDA-Q's capabilities for such applications.

I am happy to provide additional details about QTFPred's implementation, share benchmark results, or assist with testing if that would be helpful.

Best regards,
Taichi Matsubara
First Author, QTFPred
Division of Biomedical Information Analysis
Medical Institute of Bioregulation, Kyushu University

**Comment by aminx-hue** [comment]
@TaichiMatsubara Thank you for your detailed description. We will take this input into consideration when we implement this feature. It is schedule for a coming release.  Tracked in CUDAQ-294
